### DISCUSSION TITLE
Publishing failed :| Failed to publish to App Store Connect.

### DISCUSSION THREAD

**Discussion by Splagov** [codemagic-ci-cd/codemagic-docs#1171] [Q&A] [answered]
"status":"409","detail":"The bundle version must be higher than the previously uploaded version."

**No matter what TAG or branch I try to build.** 
In build settings I have auto increment. In build log I see correct next build number. 
But whet publish turn, always publish old version and old build number. And it give mentioned error. 

But what is the reason that publish old version.

In my TestFlight I have 4 builds. Next should be 5.

Build log
------
flutter build ipa --release --export-options-plist /Users/builder/export_options.plist --build-number=5

Publish log
-------
Distribution type: App Store
Min os version: 12.0
Provisioned devices: N/A
Provisions all devices: No
Supported platforms: iPhoneOS
Version code: 4
Version: 0.1.3

Has anybody experienced this issue?

**Comment by himesh-cm** [comment]
Hi @Splagov, Can you please share the build id?

**Reply by Splagov** [reply]
637734161063f408840687c8

**Reply by himesh-cm** [reply]
Hi @Splagov  - Thanks for sharing the build info, i will check and get back to you.

**Reply by himesh-cm** [reply]
Hi @Splagov,
The error from the logs are
```
The bundle version must be higher than the previously uploaded version
```

please check this documentation link on how you can set auto build versioning [here](https://docs.codemagic.io/knowledge-codemagic/build-versioning/)

**Reply by Splagov** [reply]
But I set manually build number --build-number=5

**Reply by himesh-cm** [reply]
What build number is live on the app store?

**Reply by Splagov** [reply]
4

**Reply by Splagov** [reply]
looks like it is build every time just first commit from repository. Or something.

**Reply by himesh-cm** [reply]
In that case you need to specify both - 
`--build-name=1.0.$PROJECT_BUILD_NUMBER --build-number=$PROJECT_BUILD_NUMBER`

**Reply by himesh-cm** [reply]
It is advisable to setup auto build versioning, you can check our video tutorial [here](https://www.youtube.com/watch?v=UezlgmCZLcU&t=197s)

**Reply by Splagov** [reply]
Did as you said

Build log
> flutter build ipa --release --export-options-plist /Users/builder/export_options.plist --build-name=1.0.12 --build-number=12

Publish log
Version code: 4
Version: 0.1.3

Nothing changed

**Reply by himesh-cm** [reply]
What Version code & Version is specified in your `pubspec.yaml` file?

**Reply by Splagov** [reply]
version: 0.1.5

**Reply by himesh-cm** [reply]
okay, please comment that line from the yaml file and try again.

**Reply by Splagov** [reply]
Build
> flutter build ipa --release --export-options-plist /Users/builder/export_options.plist --build-name=1.0.13 --build-number=13

Publish
>Version code: 4
>Version: 0.1.3

**Reply by himesh-cm** [reply]
Please open iOS project in Xcode and share the screenshot of the general tab of your project, which displays the build name and number

**Reply by Splagov** [reply]
![image](https://user-images.githubusercontent.com/17015762/203268204-17d9baeb-a4d2-4b8a-8f22-56105684f868.png)

**Reply by himesh-cm** [reply]
also your pubspec.yaml screenshot please

**Reply by Splagov** [reply]
![image](https://user-images.githubusercontent.com/17015762/203269461-d146d608-f38c-486f-95cb-e1b6c6a70c3b.png)

**Comment by himesh-cm** [comment]
Let's do this, please.
Add the following script under your Pre-build script in the Flutter workflow editor:
```
  scripts:
    - name: Get the latest build number
      script: | 
        LATEST_BUILD_NUMBER=$(app-store-connect get-latest-app-store-build-number $APP_APPLE_ID)
        cd ./ios # avgtool must run in the folder where xcodeproj file is located
        agvtool new-version -all $(($LATEST_BUILD_NUMBER + 1))

```
$APP_APPLE_ID is your app id from the app store account.
You can leave the build arguments blank in the text field in your workflow editor.

**Reply by Splagov** [reply]
Weird.
I have 2 applications with exactly the same settings. One of them works like a charm.

--
Pre build scripts error

Did not find any Builds matching specified filters: *.
Setting version of project Runner to: 
    1.

Also setting CFBundleVersion key (assuming it exists)

Updating CFBundleVersion in Info.plist(s)...

Updated CFBundleVersion in "Runner.xcodeproj/../Runner/Info.plist" to 1

**Reply by himesh-cm** [reply]
Is the app id correct?

**Reply by Splagov** [reply]
Something changed 

now publish log 

Version code: 1
Version: 0.1.3

**Reply by Splagov** [reply]
App Information -> An automatically generated ID assigned to your app.
Correct

**Reply by himesh-cm** [reply]
`--build-number=$(($(app-store-connect get-latest-testflight-build-number $APP_APPLE_ID) + 1))`

please add $APP_APPLE_ID as env var

**Reply by Splagov** [reply]
did at start

**Reply by himesh-cm** [reply]
How did it go?

**Reply by Splagov** [reply]
another app in my account work fine.
Every single build.

**Reply by himesh-cm** [reply]
Can you run the build again after the change of settings.

**Reply by Splagov** [reply]
what about google meet host call?
Is that possible?

---
Just changed settings.
Same
Version code: 4
Version: 0.1.3

**Reply by himesh-cm** [reply]
I can do it later today around 6 pm IST. I will drop a message here.

**Reply by Splagov** [reply]
in 2.5 hours. OK
I'll set alarm.

**Reply by Splagov** [reply]
ready

**Reply by himesh-cm** [reply]
Please share a google meet link.

**Comment by himesh-cm** [comment, marked as answer]
Hi @Splagov,
Please ensure that the plist file holds the following keys and their corresponding strings.

```
<key>CFBundleShortVersionString</key>
	<string>$(FLUTTER_BUILD_NAME)</string>
<key>CFBundleVersion</key>
	<string>$(FLUTTER_BUILD_NUMBER)</string>
```

and re-add `--build-name=1.0.0 --build-number=$(($(app-store-connect get-latest-testflight-build-number $APP_ID) + 1))` to the build arguments.

Let us know how it goes.
Thank you.

**Reply by Splagov** [reply]
So I had this one.

<key>CFBundleShortVersionString</key>
	<string>$(MARKETING_VERSION)</string>
<key>CFBundleVersion</key>
	<string>$(CURRENT_PROJECT_VERSION)</string>
	
Replaced as you suggested. Now it works.
Magic.
Thanks.

**Comment by devshally** [comment]
Hello everyone, I tried some many times to get the script to work but it just doesn't work. 

Here's my script:-

scripts:
    - name: Get the latest build number
      script: | 
        LATEST_BUILD_NUMBER=$(app-store-connect get-latest-testflight-build-number '16448XXXXX')
        cd ./ios # avgtool must run in the folder where xcodeproj file is located
        agvtool new-version -all $NEXT_BUILD_NUMBER

Here's the error I get:-

/var/folders/w2/rrf5p87d1bbfyphxc7jdnyvh0000gn/T/build_script_5_1_d8ps23a0: line 3: scripts:: command not found
/var/folders/w2/rrf5p87d1bbfyphxc7jdnyvh0000gn/T/build_script_5_1_d8ps23a0: line 4: -: command not found
/var/folders/w2/rrf5p87d1bbfyphxc7jdnyvh0000gn/T/build_script_5_1_d8ps23a0: line 5: script:: command not found
Found build number 1 from TestFlight version 1.0.40
There are no Xcode project files in this directory.  agvtool needs a project to operate.


Build failed :|
Step 5 script `Pre-build script` exited with status code 3

**Reply by himesh-cm** [reply]
Hi @devshally - can you please open another case for discussion and share the details with your build id so I can look at it?

**Reply by devshally** [reply]
Alright @himesh-cm , thank you!
